### ISSUE TITLE
Dangling mount after upgrade to 0.6.4

### ISSUE THREAD

**Issue by jpecar** [zfsonlinux/zfs#3334]
I have a small pool set aside for redundant data (repo mirrors, iso images) with deduplication enabled. It was working just fine since 0.6.1 but after upgrade I noticed that one of the mountpoints is empty. 

```
# zfs list
NAME             USED  AVAIL  REFER  MOUNTPOINT
zfsdedup         422G   162G    30K  /zfsdedup
zfsdedup/ISOs   88.9G   162G  88.9G  /export/ISOs
zfsdedup/repos   332G   162G   305G  /srv/repos


# ls -l /export/ISOs/
total 0
(should be about 100 of iso files)
```

/srv/repos works just fine.

Brief research shows:

```
# zfs unmount zfsdedup/ISOs
umount: /export/ISOs: not mounted
cannot unmount '/export/ISOs': umount failed
# zfs mount zfsdedup/ISOs
cannot mount 'zfsdedup/ISOs': filesystem already mounted
```

and

```
# grep ISO /proc/mounts 
zfsdedup/ISOs /export/ISOs zfs rw,seclabel,noatime,xattr,noacl 0 0
# umount /export/ISOs/
umount: /export/ISOs: not mounted
```

So kernel thinks something is mounted, while obviously it is not.

Any hints how to proceed from here? Reboot is not a preferred option for this machine (need to schedule downtime etc).

Might this be related to #3300 ?

**Comment by behlendorf** [comment]
It sounds like somehow the `/etc/mtab` file has gotten out of sync with `/proc/mounts`.  My first suggestion would be to try symlinking `/etc/mtab -> /proc/mounts` and then issuing the unmount.  This may work because the utilities consult the `/etc/mtab` file to determine what is and isn't mounted.

**Comment by jpecar** [comment]
Nope ... on el7 /etc/mtab is already a symlink to /proc/self/mounts, as is /proc/mounts. All /proc/pid/mounts files show ISOs as mounted.

**Comment by florianernst** [comment]
Same here.

Starting with 0.6.4 I seem to have lost some mounts. That is to say, they are still listed, but I cannot access their contents:

```
:) root@fernst:/etc$ zfs list -t filesystem -r zfspool/LXC
NAME                     USED  AVAIL  REFER  MOUNTPOINT
zfspool/LXC             30.5G   141G   228K  /var/lib/lxc
zfspool/LXC/elk         24.3G  1.65G  6.35G  /var/lib/lxc/elk
zfspool/LXC/homeserver  5.68G   143G   956M  /var/lib/lxc/homeserver
:) root@fernst:/etc$ ls -al /var/lib/lxc
total 17
drwxr-xr-x   2 root root   2 Apr 26 12:43 .
drwxr-xr-x 146 root root 148 Jun 24 09:34 ..
:) root@fernst:/etc$ df -hT /var/lib/lxc
Filesystem     Type  Size  Used Avail Use% Mounted on
zfspool/LXC    zfs   192G   21G  171G  11% /var/lib/lxc
:) root@fernst:/etc$ ls -al /etc/mtab
lrwxrwxrwx 1 root root 12 Jun 28 11:20 /etc/mtab -> /proc/mounts
:) root@fernst:/etc$ grep LXC /proc/mounts 
zfspool/LXC /var/lib/lxc zfs rw,relatime,xattr,noacl 0 0
zfspool/LXC/elk /var/lib/lxc/elk zfs rw,relatime,xattr,noacl 0 0
zfspool/LXC/homeserver /var/lib/lxc/homeserver zfs rw,relatime,xattr,noacl 0 0
```

So, still listed, still identified by the kernel, but no data accessible. And no changes possible:

```
:) root@fernst:/etc$ zfs set mountpoint=/test zfspool/LXC
umount: /var/lib/lxc/homeserver: mountpoint not found
cannot unmount '/var/lib/lxc/homeserver': umount failed
```

With 0.6.4.1 setting the ZFS module parameter `zfs_autoimport_disable=0` actually helped (I found this in #3300 which had affected me as well). Not so with 0.6.4.2 anymore.

This is on a current Debian Stretch/Sid. Please advise on how to best debug this, I'm at liberty to test at my leisure on this system.

JFTR, these are the datasets' properties:

```
:) root@fernst:/etc$ zfs get -s local,inherited all zfspool/LXC zfspool/LXC/elk zfspool/LXC/homeserver
NAME                    PROPERTY              VALUE                    SOURCE
zfspool/LXC             mountpoint            /var/lib/lxc             local
zfspool/LXC             compression           lz4                      inherited from zfspool
zfspool/LXC             relatime              on                       inherited from zfspool
zfspool/LXC/elk         mountpoint            /var/lib/lxc/elk         inherited from zfspool/LXC
zfspool/LXC/elk         compression           lz4                      inherited from zfspool
zfspool/LXC/elk         refquota              8G                       local
zfspool/LXC/elk         refreservation        8G                       local
zfspool/LXC/elk         relatime              on                       inherited from zfspool
zfspool/LXC/homeserver  mountpoint            /var/lib/lxc/homeserver  inherited from zfspool/LXC
zfspool/LXC/homeserver  compression           lz4                      inherited from zfspool
zfspool/LXC/homeserver  refquota              none                     local
zfspool/LXC/homeserver  refreservation        2G                       local
zfspool/LXC/homeserver  relatime              on                       inherited from zfspool
```

**Comment by jpecar** [comment]
Coming back to this one year later ...

The system was updated and rebooted during regular yearly maintenance, but the problem persists.
We're now on: kernel 3.10.0-327.13.1.el7, ZoL 0.6.5.6-1, srcversion: 6D4A581C24AFCEC615C1511

During reboot system dropped into an emergency shell for other reasons. At that time I checked the content of /export/ISOs and saw all files there. That made me believe this issue is resolved.
However after I fixed things that prevented the machine to fully boot and rebooted, I again found /export/ISOs to be empty.

If I could at least get rid of this dataset, that would be a step forward. So I played with mtab a bit:

    # ls -l /etc/mtab
    lrwxrwxrwx. 1 root root 17 Apr  6 09:58 /etc/mtab -> /proc/self/mounts
    # zfs get mounted zfsdedup/ISOs 
    NAME           PROPERTY  VALUE    SOURCE
    zfsdedup/ISOs  mounted   yes      -
    # rm /etc/mtab
    # grep -v ISOs /proc/self/mounts > /etc/mtab
    # zfs get mounted zfsdedup/ISOs 
    NAME           PROPERTY  VALUE    SOURCE
    zfsdedup/ISOs  mounted   no       -

But still

    # zfs destroy zfsdedup/ISOs
    cannot destroy 'zfsdedup/ISOs': dataset is busy

Any ideas on how to proceed?

**Comment by jpecar** [comment]
I was thinking about this a bit more and I have a hypotesis. I did see files in /export/ISOs in the half-booted state because /export was not yet mounted. It appears I have the mount order wrong at boot: zfs mounts /export/ISOs and then later in the boot process /export gets mounted OVER that. And conveniently, it includes empty /export/ISOs folder.

Now to figure out how to solve this ...

**Comment by freman** [comment]
Hi

A bit of a necropost, but I've just experienced and resolved a similar issue

My issue was zfs_alpha/volume2/backups was being mounted to /home/freman/backups before /home was mounted. This meant /home/freman/backups was empty and zfs refused to work with the mount/volume in any way shape or form but the mount was showing it mounted.

I had to first unmount /home, then zfs unmount zfs_alpha/volume2/backups before remounting /home and zfs mount zfs_alpha/volume2/backups

I hope this might help?
